### PR TITLE
fix #31803, covariance handling in `fieldtype_tfunc`

### LINKED ISSUE #31803: Fatal error in type inference in StatsModels
Running the tests for [Survival](https://github.com/JuliaStats/Survival.jl) on Julia nightly, we get the following error:
```
Cox: Error During Test at /home/travis/build/JuliaStats/Survival.jl/test/runtests.jl:188
  Got exception outside of a @test
  fatal error in type inference (type bound)
  Stacktrace:
   [1] StatsModels.Terms(::Formula) at /home/travis/.julia/packages/StatsModels/pBxdt/src/formula.jl:341
   [2] #fit#36(::Dict{Any,Any}, ::Base.Iterators.Pairs{Symbol,Float64,Tuple{Symbol},NamedTuple{(:tol,),Tuple{Float64}}}, ::typeof(fit), ::Type{CoxModel}, ::Formula, ::DataFrame) at /home/travis/.julia/packages/StatsModels/pBxdt/src/statsmodel.jl:66
   [3] (::getfield(StatsBase, Symbol("#kw##fit")))(::NamedTuple{(:tol,),Tuple{Float64}}, ::typeof(fit), ::Type{CoxModel}, ::Formula, ::DataFrame) at ./none:0
   [4] #coxph#20(::Base.Iterators.Pairs{Symbol,Float64,Tuple{Symbol},NamedTuple{(:tol,),Tuple{Float64}}}, ::typeof(coxph), ::Formula, ::DataFrame) at /home/travis/build/JuliaStats/Survival.jl/src/cox.jl:203
   [5] (::getfield(Survival, Symbol("#kw##coxph")))(::NamedTuple{(:tol,),Tuple{Float64}}, ::typeof(coxph), ::Formula, ::DataFrame) at ./none:0
   [6] top-level scope at /home/travis/build/JuliaStats/Survival.jl/test/runtests.jl:192
   [7] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Test/src/Test.jl:1113
   [8] top-level scope at /home/travis/build/JuliaStats/Survival.jl/test/runtests.jl:189
   [9] include at ./boot.jl:328 [inlined]
   [10] include_relative(::Module, ::String) at ./loading.jl:1094
   [11] include(::Module, ::String) at ./Base.jl:31
   [12] include(::String) at ./client.jl:431
   [13] top-level scope at none:5
   [14] eval(::Module, ::Any) at ./boot.jl:330
   [15] exec_options(::Base.JLOptions) at ./client.jl:271
   [16] _start() at ./client.jl:464
```
The line in StatsModels identified by the stack trace is https://github.com/JuliaStats/StatsModels.jl/blob/v0.5.0/src/formula.jl#L341.

This does not occur on 0.7, 1.0, or 1.1, so it seems like a regression.

### PR DESCRIPTION
**Pull request by JeffBezanson**
fixes #31803

### CODE REVIEW

File: base/compiler/tfuncs.jl [hunk 1]
```diff
@@ -939,7 +939,7 @@ function _fieldtype_tfunc(@nospecialize(s), exact::Bool, @nospecialize(name))
         ft = ftypes[fld]
     end
 
-    exactft = exact || !has_free_typevars(ft)
+    exactft = exact || (!has_free_typevars(ft) && u.name !== Tuple.name)
```
Comment (Keno) [inline base/compiler/tfuncs.jl]: Doesn't seem specific to the typevar case?
```
julia> Core.Compiler.fieldtype_tfunc(Type{Tuple{Any}}, Const(1))
Const(Any, false)
```

Comment (JeffBezanson) [reply base/compiler/tfuncs.jl]: `fieldtype(Tuple{Any}, 1)` does return `Any`, so I think that result is ok.

Comment (Keno) [reply base/compiler/tfuncs.jl]: ah, right

**Comment by vtjnash** [comment]
I suggest also applying the same fix on line 909:
```suggestion
    exactft1 = exact || (!has_free_typevars(ft1) && u.name !== Tuple.name)
 ```

**Review by vtjnash — APPROVED** [review]

**Pull request merged**